src/mm-input/mm-input.js

```diff
@@ -4,213 +4,175 @@
-		if(this.icon === "search" && newVal) {
-			this.paddingRight = this.PADDING_ICON;
-			this.$.icon.style.display = "block";
-		}
-	},
-
-	iconChanged: function(oldVal, newVal) {
-		if(this.icon !== "search" && newVal) {
-			this.paddingRight = this.PADDING_ICON;
-			this.$.icon.style.display = "block";
-		}
-	},
+		},
 
-	clearChanged: function(oldVal, newVal) {
-		this.paddingRight = this.PADDING_ICON;
-	},
-	
-	valueChanged: function(oldVal, newVal) {
-		//update model if needed
-		this.updateModel();
-		this.fire("changed", { value: this.value });
-		
-		if (!this.autoItemSelected){
-			// the user is typing
-			this.validate();
-			this.updateUI();
-		} else {
-			// the user selected an item from the autocomplete list
-			this.autoItemSelected = false;
-		}
-	},
-
-	inputWidthChanged: function (oldVal, newVal) {
-		// set close panel width if the input width changed:
-		if (this.$ && this.$.closePanel) {
-			if (this.autocomplete && this.overflow === "visible") {
-				this.$.closePanel.style.width = "auto";
-				this.$.closePanel.style.minWidth = newVal + "px";
-			} else {
-				this.$.closePanel.style.width = newVal + "px";
+		_typeChanged: function(newVal, oldVal) {
+			var types = /(text|password|email|number|tel|search|url)/ig;
+			if (!types.test(newVal)) {
+				this.type = "text";
 			}
-		}
-	},
-
-	validate: function() {
-		var valid = this.validator();
-
-		if (valid) {
-			this.error = false;
-		} else {
-			this.error = true;
-		}
-	},
-
-	validator: function() {
-		switch(this.validation) {
-			case "empty":
-				return this.value.length > 0;
-			case "int":
-				return this.inter.test(this.value);
-			case "decimal":
-				return this.floater.test(this.value);
-			case "email":
-				return this.email.test(this.value);
-			case "alpha":
-				return this.alpha.test(this.value);
-			default:
-				return true;
-		}
-	},
+		},
 
-	updateUI: function() {
-		var dataFilter = function(a) {
-			return a.name.indexOf(this.value) != -1;
-		};
-
-		//handle the clear and search state styling:
-		if (this.value && this.value.length > 0) {
-			if (this.clear) {
-				this.$.clearBtn.style.display = "block";
-			}
-			if (this.icon === "search" && this.search && this.clear || this.icon && this.clear) {
-				this.$.icon.style.display = "none";
-			}
-			if (this.autocomplete) {
-				this.searchItems = this.data.filter(dataFilter.bind(this));
-				if (this.searchItems.length > 0) {
-					this.open();
-				}
-				// if (this.searchItems.length > this.maxItems) {
-				// this.searchItems = this.searchItems.splice(0, this.maxItems);
-				// }
+		_searchChanged: function(newVal, oldVal) {
+			if (newVal) { 
+				this.icon = "search";
+			} else if (!newVal && this.icon === "search") {
+				this.icon = false;
 			}
-		} else {
-			if (this.clear) {
-				this.$.clearBtn.style.display = "none";
-			}
-			if (this.icon === "search" && this.search && this.clear || this.icon && this.clear) {
+		},
+
+		_iconChanged: function(newVal, oldVal) {
```

Comment:
per @derekolson's comment on friday it makes more sense if these are being done declaratively in the template and for internal usage we can use the 'hidden' attribute ala ```hidden$="{{_updateIcon(icon)}}"```